### PR TITLE
Lower fetch priority of metrics

### PR DESCRIPTION
**Pull request by filiptronicek**
## Description

Gives our metrics calls less priority in comparison to mission-critical resources like API requests. 

## How to test

1. Make sure things in the [preview environment](https://ft-metrics3af7f1b94e.preview.gitpod-dev.com/workspaces) still work

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

### CODE REVIEW

**Comment by filiptronicek** [comment]
Seems like prioritization is properly applied. Please test in Chrome :)

image

File: components/supervisor/frontend/src/ide/ide-metrics-service-client.ts [hunk 1]
```diff
@@ -70,6 +70,7 @@ export class IDEMetricsServiceClient {
                 method: "POST",
                 body: JSON.stringify(params),
                 credentials: "omit",
+                priority: "low",
```
Comment (kylos101) [inline components/supervisor/frontend/src/ide/ide-metrics-service-client.ts:73]: @filiptronicek are you able to create the scenario where you can see metrics stall / get stuck in Pending?

Comment (filiptronicek) [reply components/supervisor/frontend/src/ide/ide-metrics-service-client.ts:73]: nope, and not even with forced HTTP/1.1. Theoretically, though, if two requests were made at the exact same time in the dashboard now (for instance, one for starting a workspace or watching its status), that one would get prioritized, making us less prone to issues with the 6 concurrent HTTP/1.1 TCP connections per origin.

Comment (kylos101) [reply components/supervisor/frontend/src/ide/ide-metrics-service-client.ts:73]: Will priority work with HTTP/1/1? I see references to HTTP/2 and HTTP/3 [here](https://imkev.dev/fetchpriority-opportunity#fetchpriority-all-the-things).

It looks like it is broadly supported, [except in Firefox](https://developer.mozilla.org/en-US/docs/Web/API/HTMLScriptElement/fetchPriority).

Can GPT drum up a simple set of code to help test whether this works? For example, to make an artificially high number of calls in two different groups (a high and a low)?

Comment (filiptronicek) [reply components/supervisor/frontend/src/ide/ide-metrics-service-client.ts:73]: > Will priority work with HTTP/1/1? I see references to HTTP/2 and HTTP/3 [here](https://imkev.dev/fetchpriority-opportunity#fetchpriority-all-the-things).

I don't believe this to be dependent on the HTTP version. The linked article just refers to h2 and h3 in relation to the data from HTTP Archive. 

> It looks like it is broadly supported, [except in Firefox](https://developer.mozilla.org/en-US/docs/Web/API/HTMLScriptElement/fetchPriority).

Yes, Firefox thankfully just ignores it, so there's no need for any special handling.

> Can GPT drum up a simple set of code to help test whether this works? For example, to make an artificially high number of calls in two different groups (a high and a low)?

I'll take a look.

Comment (filiptronicek) [reply components/supervisor/frontend/src/ide/ide-metrics-service-client.ts:73]: I have gotten some pretty ok results when firing 500 requests at once with [this fiddle](https://jsfiddle.net/hyL8pefk/):
```
#1
"Mean response time for priority high: 2001.64 ms"
"Mean response time for priority low: 2128.59 ms"

#2
"Mean response time for priority high: 2001.64 ms"
"Mean response time for priority low: 2128.59 ms"
```

Comment (kylos101) [reply components/supervisor/frontend/src/ide/ide-metrics-service-client.ts:73]: I tested while on HTTP/1.1, the low priority group always takes 500ms longer. 👍 

[Here's an updated fiddle](https://jsfiddle.net/7s59mzrg/) that hits an endpoint and introduces some delay. It's really slow with HTTP/1.1. But, it shows high priority is "much better".

Results for updated fiddle:
☁️ "Running fiddle"
"Priority high: Mean response time: 43465.43 ms, Count: 46"
"Priority low: Mean response time: 125647.33 ms, Count: 54"

File: components/dashboard/src/start/StartWorkspace.tsx [hunk 2]
```diff
@@ -55,7 +55,7 @@ export function parseProps(workspaceId: string, search?: string): StartWorkspace
     const runsInIFrame = window.top !== window.self;
     return {
         workspaceId,
-        runsInIFrame: window.top !== window.self,
+        runsInIFrame,
```
Comment (kylos101) [inline components/dashboard/src/start/StartWorkspace.tsx:58]: 👋 hey @filiptronicek, why is this change necessary?

File: components/gitpod-protocol/package.json [hunk 3]
```diff
@@ -32,7 +32,7 @@
         "mocha": "^10.2.0",
         "rimraf": "^2.6.2",
         "ts-node": "^10.4.0",
-        "typescript": "~4.4.2",
+        "typescript": "^5.5.4",
```
Comment (kylos101) [inline components/gitpod-protocol/package.json:35]: Hey @filiptronicek , this seems like a big jump, which could have unintended consequences. May I ask, is this required to support priority? Or is it a drive by change to update typescript?

Comment (filiptronicek) [reply components/gitpod-protocol/package.json:35]: Yes, typescript on this version does not have the DOM type for `priority` in `RequestInit`.

File: components/dashboard/src/start/StartWorkspace.tsx (hunk 2 above)
Comment (filiptronicek) [reply components/dashboard/src/start/StartWorkspace.tsx:58]: No necessary, just reuses a variable we already had for simplicity. Happy to revert as it's unrelated to the change itself and I just happened to make it while investigating workspace starts.

Comment (kylos101) [reply components/dashboard/src/start/StartWorkspace.tsx:58]: Oh, I see. 

It seems safe to keep scope wise, no need to revert (I should have looked above earlier).

**Review by kylos101 — APPROVED** [review]
Code makes sense, I tested with HTTP/1.1, it works really well!

image


Also, I adjusted the fiddle to hit an endpoint that introduces delay, and low priority definitely takes a back seat to high priority.

File: components/supervisor/frontend/src/ide/ide-metrics-service-client.ts (hunk 1 above)
Comment (filiptronicek) [reply components/supervisor/frontend/src/ide/ide-metrics-service-client.ts:73]: I think we're not going to miss out on them, rather just let them queue to a later time. I'm happy to take a look when this hits production
